packages/grid/src/vaadin-grid-selection-mixin.js

```diff
@@ -68,19 +77,34 @@ export const SelectionMixin = (superClass) =>
      * @protected
      */
     _toggleItem(item) {
-      const index = this._getItemIndexInArray(item, this.selectedItems);
-      if (index === -1) {
+      if (!this._isSelected(item)) {
         this.selectItem(item);
       } else {
         this.deselectItem(item);
       }
     }
 
+    /** @private */
+    _itemIdPathChanged() {
+      this.__cacheSelectedKeys();
+    }
+
     /** @private */
     _selectedItemsChanged() {
+      this.__cacheSelectedKeys();
       this.requestContentUpdate();
     }
 
+    /** @private */
+    __cacheSelectedKeys() {
+      if (this.selectedItems) {
```

Comment:
I'm not sure if we have a scenario where `selectedItems` can be undefined or null, but in that case the `__selectedKeys` would become stale.

As an alternative, the condition could be replaced with:
```js
const selectedItems = this.selectedItems || [];
```

Reply:
This method is essentially a copy/paste of `__cacheExpandedKeys`; I didn't give it much thought. I'll adapt `__cacheSelectedKeys` per your suggestion.